
<a href="https://colab.research.google.com/github/Mhshohag2256/CSC-448/blob/main/Shohag_Mehedi_A4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Mehedi H Shohag

In [ ]:
# Imports and pip installations (if needed)
import numpy as np
import pandas as pd
from sklearn import svm, datasets
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')

# Part 1: Load the dataset

In [ ]:
# Load the dataset (load remotely, not locally)
iris = datasets.load_iris()

# Output the first 15 rows of the data
iris_df = pd.DataFrame(iris['data'], columns = iris.feature_names)
iris_df['species'] = iris.target

# Display a summary of the table information (number of datapoints, etc.)
print(iris_df.info())
print('Total Data Points for each spices:',len(iris_df))
print('All together datapoints :',iris_df.size)
iris_df.head(15)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   species            150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB
None
Total Data Points for each spices: 150
All together datapoints : 750


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    species  
0         0  
1         0  
2         0  
3         0  
4         0  
5         0  
6         0  
7         0  
8         0  
9         0  
10        0  
11        0  
12        0  
13        0  
14        0

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes? <br>

The data set includes 150 sets of data in 4 categories which are 'Sepal Length','Sepal Width','Petal Length' and, 'Petal Width'.And there are in total 3 species which are 'Iris setosa', 'Iris virginica' and 'Iris versicolor'. Each of the species has 50 data points. <br>

My features would be the physical properties of the species which are 'Sepal Length','Sepal Width','Petal Length' and, 'Petal Width' and label would be 3 different species.

The mapping would be based on the features length value, whichever species matches.


# Part 2: Split the dataset into train and test

In [ ]:
# Take the dataset and split it into our features (X) and label (y)
X = iris.data
y = iris.target
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, train_size=0.9)

# Part 3: Logistic Regression

In [ ]:
# i. Use sklearn to train a LogisticRegression model on the training set
logreg = LogisticRegression(C=1e5,max_iter = 500)
logreg.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
print(pd.DataFrame(logreg.predict_proba([[4.9,3.1,1.5,0.1]]), columns = ['Setosa', 'Versicolor', 'Virginica']))
print('Setosa 100% and the correct prediction is Setosa.')
print('-------------------------------------------------')
# iii. Report on the score for Logistic regression model, what does the score measure?
sc = StandardScaler()
sc.fit(X_train)
X_test_std = sc.transform(X_test)
print('Score :',logreg.score(X_test_std, y_test))

   Setosa    Versicolor     Virginica
0     1.0  8.649969e-08  1.110241e-32
Setosa 100% and the correct prediction is Setosa.
-------------------------------------------------
Score : 0.5333333333333333


The score is 53%. It means out of the 10% datasets it predicted 53% of the datapoints accurately and the rest (47%) predicted wrong.

In [ ]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
print('Coefficents :',logreg.coef_)
print('-----------------------------------------------------')
print('intercepts :',logreg.intercept_)

Coefficents : [[  3.81778207   9.09678725 -12.22222575  -5.72269387]
 [ -0.65402157  -1.26762577   1.61679342  -6.10085876]
 [ -3.16376051  -7.82916148  10.60543233  11.82355263]]
-----------------------------------------------------
intercepts : [  1.94552558  19.00320662 -20.9487322 ]


# Part 4: Support Vector Machine

In [ ]:
# i. Use sklearn to train a Support Vector Classifier on the training set
SVM = svm.SVC(kernel='linear')
SVM.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
print(pd.DataFrame(logreg.predict_proba([[4.9, 2.4, 3.3, 1.]]), columns = ['Setosa', 'Versicolor', 'Virginica']))
print('Versicolor 100% and the correct prediction is Versicolor.')
print('---------------------------------------------------------')

# iii. Report on the score for the SVM, what does the score measure?
SVM.score(X_test_std, y_test)

         Setosa  Versicolor     Virginica
0  1.744111e-07         1.0  1.367303e-09
Versicolor 100% and the correct prediction is Versicolor.
---------------------------------------------------------


0.4

The score is 40%. It means out of the 10% datasets it predicted 40% of the datapoints accurately and the rest (60%) predicted wrong.

# Part 5: Neural Network

In [ ]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
print('Configuration 1')
NN = MLPClassifier(max_iter = 1000, activation='relu', solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(16,), random_state=1)
NN.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
print(pd.DataFrame(NN.predict_proba([[7.7, 2.8, 6.7, 2.]]), columns = ['Setosa', 'Versicolor', 'Virginica']))
print('Virginica 100% and the correct prediction is Virginica.')
print('---------------------------------------------------------')

# iii. Report on the score for the Neural Network, what does the score measure?
print(NN.score(X_test_std, y_test))

Configuration 1
          Setosa    Versicolor  Virginica
0  1.768109e-134  9.885426e-09        1.0
Virginica 100% and the correct prediction is Virginica.
---------------------------------------------------------
0.6


The score is 60%. It means out of the 10% datasets it predicted 60% of the datapoints accurately and the rest (40%) predicted wrong.

In [ ]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)
print('Configuration 2')
NN = MLPClassifier(max_iter = 2000, activation='tanh', solver='adam', alpha=1e-5, hidden_layer_sizes=(32,), random_state=1)
NN.fit(X_train, y_train)
print(NN.score(X_test_std, y_test))
print('---------------------------------------------------------')

print('Configuration 3')
NN = MLPClassifier(max_iter = 500, activation='relu', solver='adam', alpha=1e-5, hidden_layer_sizes=(64,), random_state=1)
NN.fit(X_train, y_train)
print(NN.score(X_test_std, y_test))
print('---------------------------------------------------------')

print('Configuration 4')
NN = MLPClassifier(max_iter = 1000, activation='relu', solver='sgd', alpha=1e-5, hidden_layer_sizes=(64,), random_state=1)
NN.fit(X_train, y_train)
print(NN.score(X_test_std, y_test))
print('---------------------------------------------------------')

print('Configuration 5')
NN = MLPClassifier(max_iter = 800, activation='logistic', solver='sgd', alpha=1e-5, hidden_layer_sizes=(16,), random_state=1)
NN.fit(X_train, y_train)
print(NN.score(X_test_std, y_test))


Configuration 2
0.6666666666666666
---------------------------------------------------------
Configuration 3
0.8666666666666667
---------------------------------------------------------
Configuration 4
0.8
---------------------------------------------------------
Configuration 5
0.7333333333333333


In my configuration 3 I've used activation function relu and optimizer adam with 64 hidden layer, and I've got the highest score like you professor which is .8666 😉

# Part 6: K-Nearest Neighbors

In [ ]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

KNN = KNeighborsClassifier(n_neighbors=5)
KNN.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
print(pd.DataFrame(KNN.predict_proba([[7.7, 2.8, 6.7, 2.]]), columns = ['Setosa', 'Versicolor', 'Virginica']))
print('Virginica 100% and the correct prediction is Virginica.')
print('---------------------------------------------------------')

# iii. Report on the score for kNN, what does the score measure?
KNN.score(X_test_std, y_test)

   Setosa  Versicolor  Virginica
0     0.0         0.0        1.0
Virginica 100% and the correct prediction is Virginica.
---------------------------------------------------------


0.4

The score is 40%. It means out of the 10% datasets it predicted 40% of the datapoints accurately and the rest (60%) predicted wrong.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise? <br>

Logistic Regression Accuracy = 53% <br>
Support Vector Machine Accuracy = 40% <br>
Neural Network Accuracy(best) = 86% <br>
K-Nearest Neighbors Accuracy = 40% <br>

Out of all the models neural network performs very well and gave 86% accuracy. The reason I think, in neural network model architecture there are multiple hidden layers and activation function, loss fuction, and optimizers comes into play and passing the data over and over the model learns automatically and corrects itself and that's how I think it gets a good accuracy. <br>
Things that surprised me is using this model it knows which iris species is which like human. Machine started learning like human. Which is fascinating.